### PR TITLE
Refactor the event channel

### LINKED ISSUE #1909: [Ballista] Refactor the event channel
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently there're many places leveraging event channel for event-based processing. Many parts can be reused.

**Describe the solution you'd like**

It's possible to extract common things.
- EventAction<E> for common trait
- EventLoop<E> for common struct

**Additional context**
#1704

### PR DESCRIPTION
**Pull request by yahoNanJing**
# Which issue does this PR close?

Closes #1909.

 # Rationale for this change

Mentioned in #1909.

# What changes are included in this PR?

Introduce the following for event-based processing:
- EventAction for common trait
- EventLoop for common struct

# Are there any user-facing changes?

It's blocked by #1911.

### CODE REVIEW

File: ballista/rust/core/src/event_loop.rs
Comment (yahoNanJing) [reply ballista/rust/core/src/event_loop.rs:73]: Yes. In general, the channel won't be closed until the scheduler is down. The dangling task should be handled by another way, like speculative task execution.

File: ballista/rust/scheduler/src/scheduler_server/mod.rs
Comment (yahoNanJing) [reply ballista/rust/scheduler/src/scheduler_server/mod.rs:125]: Actually, in this PR the EventLoop don't need Arc. However, from the whole processing figure in #1704, the event loop will be used in somewhere else. The related PR is under progress.

Comment (thinkharderdev) [reply ballista/rust/scheduler/src/scheduler_server/mod.rs:125]: I see. In that case, does the `EventLoop` need to be passed around, or can we just pass a `Sender` to different producers? In that case, the `EventLoop` could still be owned and multiple producers could hold a clone of the event loops `Sender`. 

Maybe @alamb could comment on how much we should care to avoiding `unsafe`.

File: ballista/rust/scheduler/src/scheduler_server/event_loop.rs
Comment (alamb) [inline ballista/rust/scheduler/src/scheduler_server/event_loop.rs:120]: I am not sure what the plan is here, but sometimes the `todo!()` macro gets used in situations so the TODO isn't silently forgotten. However, since `todo!()` panic's this may not be possible if the functions are called

```suggestion
    fn on_start(&self) {
      todo!();
    }
```

File: ballista/rust/scheduler/src/scheduler_server/mod.rs
Comment (alamb) [reply ballista/rust/scheduler/src/scheduler_server/mod.rs:125]: The philosophy in Rust in general is to avoid `unsafe` as much as possible. Typically it `unsafe` is used for one of two reasons:
1. Interfacing with other languages (like `C/C++`) that don't have the same lifetime system as Rust
2. Performance where the overhead of doing things like array bounds checks is too great

Neither of those situations seems to apply here. 

I would personally suggest figuring out some way to avoid this `unsafe` block - @thinkharderdev  has one suggestion; 

I don't know this code well enough to offer anything specific, but I would say another pattern for something like this is "Interior Mutability" which is a fancy rust way of saying use something like `Mutex`. Perhaps

```rust
event_loop: Option<Arc<Mutex<EventLoop<SchedulerServerEvent>>>>,
```

**Review by alamb — COMMENTED** [review]
Thanks @yahoNanJing  -- I skimmed this PR and the basic idea of breaking the system down into smaller modules I think is a great direction.

One thing I wanted to mention that might help with review / merge speed in the future is to break such refactoring PRs into two parts:

1. A PR that just moves code around / breaks up files but doesn't change behavior
2. The PR that changes behavior

The reason is that often moving files around causes large diffs in github but if there is no change in behavior can be straightforward to review / merge

The changes in behavior may take some more thought / comment and can be harder to understand in a larger diff. 

Also, a larger PR has a greater chance to accumulate conflicts so if behavior changes get delayed in review, that can also cause issues

**Comment by liukun4515** [comment]
> Thanks @yahoNanJing -- I skimmed this PR and the basic idea of breaking the system down into smaller modules I think is a great direction.
> 
> One thing I wanted to mention that might help with review / merge speed in the future is to break such refactoring PRs into two parts:
> 
> 1. A PR that just moves code around / breaks up files but doesn't change behavior
> 2. The PR that changes behavior
> 
> The reason is that often moving files around causes large diffs in github but if there is no change in behavior can be straightforward to review / merge
> 
> The changes in behavior may take some more thought / comment and can be harder to understand in a larger diff.
> 
> Also, a larger PR has a greater chance to accumulate conflicts so if behavior changes get delayed in review, that can also cause issues

I think we can review this pull request after rebase.
@alamb

File: ballista/rust/core/src/event_loop.rs
Comment (yahoNanJing) [reply ballista/rust/core/src/event_loop.rs:73]: Thanks @alamb. I'll add the informative log.

File: ballista/rust/scheduler/src/scheduler_server/mod.rs
Comment (yahoNanJing) [reply ballista/rust/scheduler/src/scheduler_server/mod.rs:125]: Hi @alamb, the reason I prefer not to add Mutex here is that the mutation only happens in the initialization phase. In other places, only it's only for reading. And we can avoid the overhead of locks.

@thinkharderdev, this PR is for providing some encapsulation for the channel processing and extracting some common behaviors in the EventLoop.

**Comment by yahoNanJing** [comment]
> Thanks @yahoNanJing -- I skimmed this PR and the basic idea of breaking the system down into smaller modules I think is a great direction.
> 
> One thing I wanted to mention that might help with review / merge speed in the future is to break such refactoring PRs into two parts:
> 
> 1. A PR that just moves code around / breaks up files but doesn't change behavior
> 2. The PR that changes behavior
> 
> The reason is that often moving files around causes large diffs in github but if there is no change in behavior can be straightforward to review / merge
> 
> The changes in behavior may take some more thought / comment and can be harder to understand in a larger diff.
> 
> Also, a larger PR has a greater chance to accumulate conflicts so if behavior changes get delayed in review, that can also cause issues

Thanks @alamb. The background of doing this refactoring is mainly for achieving the final goal of #1704. Without these refactoring, all of the complex things will messed up in one or two files, which is really hard for maintenance. 

For the step by step refactoring, every PR will be tested and verified without changing the behavior.

**Comment by yahoNanJing** [comment]
> Thanks @yahoNanJing -- I skimmed this PR and the basic idea of breaking the system down into smaller modules I think is a great direction.
> 
> One thing I wanted to mention that might help with review / merge speed in the future is to break such refactoring PRs into two parts:
> 
> 1. A PR that just moves code around / breaks up files but doesn't change behavior
> 2. The PR that changes behavior
> 
> The reason is that often moving files around causes large diffs in github but if there is no change in behavior can be straightforward to review / merge
> 
> The changes in behavior may take some more thought / comment and can be harder to understand in a larger diff.
> 
> Also, a larger PR has a greater chance to accumulate conflicts so if behavior changes get delayed in review, that can also cause issues

Thanks @alamb. The background of doing this refactoring is mainly for achieving the final goal of #1704. Without these refactoring, all of the complex things will messed up in one or two files, which is really hard for maintenance. 

For the step by step refactoring, every PR will be tested and verified without changing the behavior.

File: ballista/rust/scheduler/src/scheduler_server/event_loop.rs
Comment (yahoNanJing) [reply ballista/rust/scheduler/src/scheduler_server/event_loop.rs:120]: Agree. I'll change it to be comments.

File: ballista/rust/scheduler/src/scheduler_server/mod.rs
Comment (thinkharderdev) [reply ballista/rust/scheduler/src/scheduler_server/mod.rs:125]: Right, I think it's a great idea to encapsulate the event loop, but I think we can avoid the `unsafe` and the overhead of a `Mutex` or `RwLock` by decoupling the producer and consumer interface. If we want to encapsulate the error handling around sending an event on the underlying channel then perhaps we can create an explicit `EventProducer` like:

```rust
pub struct EventProducer<E> {
  tx_event: mpsc::Sender<E>
}

impl<E> EventProducer<E> {
      pub async fn post_event(&self, event: E) -> Result<()> {
        if let Some(tx_event) = self.tx_event.as_ref() {
            tx_event.send(event).await.map_err(|e| {
                BallistaError::General(format!("Fail to send event due to {}", e))
            })?
        } else {
            warn!("Event sender not exist!!!");
        }

        Ok(())
    }
} 

impl<E> EventLoop<E> {
  pub fn producer(&self) -> Result<EventProducer<E>> {
    EventPrdocuer::new(self.tx_event.cloned().ok_or(...)?)
  }
}
```

File: ballista/rust/core/src/event_loop.rs
Comment (alamb) [inline ballista/rust/core/src/event_loop.rs:89]: 👍

File: ballista/rust/scheduler/src/scheduler_server/mod.rs
Comment (alamb) [reply ballista/rust/scheduler/src/scheduler_server/mod.rs:125]: My personal recommendation is go with the `safe` approach (even if it seems slower at first as it may require locks). My rationale is:

I have spent countless hours tracking down strange, subtle and hard to reproduce bugs related to memory corruption when working on C/C++ systems. My Rust experience has largely been free of such pain and I think the avoidance of that pain a key advantage of Rust and makes some of the painful parts of rust (like the borrow checker) worthwhile.

I also think the overhead of taking a mutex is likely to be low, and you could make it lower still by using using a `RwLock` instead. 

Even better would be if you can refactor the code so that it is clear when the initialization has occured so you need neither mutex nor unsafe, perhaps as @thinkharderdev  is suggesting

I like to think in most matters I am pretty pragmatic, but the use of `unsafe` is something I do feel quite strongly about.

**Comment by alamb** [comment]
I am happy with this PR now other than the `unsafe` -- if @thinkharderdev or @liukun4515  is ok with the `unsafe` approach, I'll be happy to merge this PR. Otherwise I really suggest going with a `RwLock`

**Comment by thinkharderdev** [comment]
> I am happy with this PR now other than the `unsafe` -- if @thinkharderdev or @liukun4515 is ok with the `unsafe` approach, I'll be happy to merge this PR. Otherwise I really suggest going with a `RwLock`

I would also prefer we remove the `unsafe` block before merging.

File: ballista/rust/scheduler/src/scheduler_server/mod.rs
Comment (yahoNanJing) [reply ballista/rust/scheduler/src/scheduler_server/mod.rs:125]: Thanks @alamb and @thinkharderdev. I think to avoid unsafe here, the only way is to use RwLock or Mutex. Although I still think it's not necessary, I will change it to use RwLock to avoid unsafe.

Comment (yahoNanJing) [reply ballista/rust/scheduler/src/scheduler_server/mod.rs:125]: Thanks @thinkharderdev and @alamb. I'll refine the code according to the @thinkharderdev's suggestion.

**Comment by yahoNanJing** [comment]
Hi @thinkharderdev and @alamb, could you help review the latest commit for fixing unsafe.

**Review by alamb — APPROVED** [review]
Looks great to me. Thank you @yahoNanJing  and @thinkharderdev 

I believe @liukun4515  is planning to review this tomorrow so I will hold off on merging until his review is complete

**Review by thinkharderdev — APPROVED** [review]
Looks great!

**Review by liukun4515 — APPROVED** [review]
LGTM

**Comment by alamb** [comment]
Thanks again @yahoNanJing

**Pull request merged**
